Feat/go add query to all listings

OlfaBensoussia opened the pull request:
## Description
We made this PR to be able to support adding a filtering query while listing assets and contract definitions.
This resolves #30 and suggests a refactored way to validate query inputs.

## Approach
- Updated the endpoints and the requests methods
-  Added a new package **apivalidator** under **internal** to avoid copy-pasting the same object and function definition in multiple files.
- Renamed `ListPoliciesInput` to `QueryInput` because we will be referring to the same object in all of the **POST** requests [Check [this link ](https://app.swaggerhub.com/apis/eclipse-edc-bot/management-api/0.0.1-SNAPSHOT#/QuerySpecDto)for reference]
- Renamed `validateQueryPoliciesInput` to `ValidateQueryInput`
- Fixed what seemed to be some typos in the returned error messages

### Open Questions and Pre-Merge TODOs
- I was thinking that we could move `ApiErrorDetail` to the **apivalidator** package too and use it for the other entities as well (assets & contract definitions).

OlfaBensoussia commented:
@NabilHouidi @AhmedGrati Tagging you guys here because I know that Youssef will be OOO until Thursday. 
 I would really appreciate it if you could take some time to review those changes and provide any feedback or suggestions for improvement.
I understand that you may be busy, so please feel free to take your time with this.

AhmedGrati reviewed (commented):
It's an LGTM for me. Great job @OlfaBensoussia 👌
Just one small comment, can we add some unit tests to the `apivalidator`?
As for the API error, @NabilHouidi opened a PR to unify errors, so we'll get rid of the API errors.

AhmedGrati commented:
@OlfaBensoussia Can you please also update the `examples/assetsCrud/main.go` file with the introduced changes?

OlfaBensoussia commented:
@AhmedGrati I added everything and updated a few things all along the way.
The git history is not the cleanest but I will make sure to take care of that once I have your approval.

AhmedGrati commented on go/common/apivalidator/validate_query.go:
```diff
@@ -0,0 +1,18 @@
+		return nil
+	}
+	if *sortOrder != SortOrderAscendant && *sortOrder != SortOrderDescendant {
+		return fmt.Errorf(ErrInvalidSortOrder,
```
@OlfaBensoussia we added a new way of handling errors and I would suggest adapting this to the existing error handling method can be better for consistency reasons.

AhmedGrati commented on go/examples/assetsCrud/main.go:
```diff
@@ -164,4 +166,23 @@ func main() {
+	filteredAssets, err := client.ListAssets(filter)
+	
+	if err != nil {
+		fmt.Printf("error while listing assets with filter %v", filter )
```
small nit: printing the error here can be useful.

AhmedGrati commented on go/examples/contractDefinitionsCrud/main.go:
```diff
@@ -93,4 +94,22 @@ func main() {
+	filteredContractDefinitions, err := client.ListContractDefinitions(filter)
+		
+	if err != nil {
+		fmt.Printf("error while listing contract definitions with filter %v", filter )
```
Same as above.

AhmedGrati commented on go/examples/policiesCrud/main.go:
```diff
@@ -56,29 +57,49 @@ func main() {
+		
 	if err != nil {
-		fmt.Printf("error while getting an policy by id %+s. Full error:\n%v", policyId, err)
+		fmt.Printf("error while listing policies with filter %v", filter )
```
same as above.

AhmedGrati commented on go/service/assets/api_op_ListAssets.go:
```diff
@@ -15,11 +17,27 @@ type AssetOutput struct {
+	var queryInput apivalidator.QueryInput
+
+	if len(args) > 0 {
+		queryInput = args[0]
```
Just a question here: 
why do we take only one `queryInput`? If the connector API accepts only one query Input and not an array, I suggest updating the `ListAssets` parameter to a `*apivalidator.QueryInput` rather than an array of `apivalidator.QueryInput`. wdyt?

OlfaBensoussia replied:
That line of code checks if we have a `queryInput` in the first place because it's optional. As you mentioned, the connector API expects only one query max.
The only reason why I went with this approach is to be able to call `ListAssets()` this way rather than `ListAssets(nil)` in case we don't have a query. [You can see pictures above of how I coded it previously]
![343574294_583671690412259_5966455409479978209_n](https://github.com/Think-iT-Labs/edc-connector-client/assets/61057075/d7e4643b-4e55-4cc5-bf61-3c1c4c67a359)

![344762782_6066875723439747_4831448852812178306_n](https://github.com/Think-iT-Labs/edc-connector-client/assets/61057075/bb4e6312-350e-4b5c-8b6d-ab9379abe828)

The approach is very biased because it follows the way we do it in Javascript 😅 But I'm open to change it too.

AhmedGrati replied:
Yes makes more sense now, I see that your approach is better now. Thank you 🙏.

AhmedGrati commented on go/service/assets/api_op_ListAssets.go:
```diff
@@ -15,11 +17,27 @@ type AssetOutput struct {
 
-	req, err := http.NewRequest(http.MethodGet, endpoint, nil)
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error while marshaling list assets query: %v", err)
```
Same as above, I highly recommend updating the error handling method to the new one. You can see an example in line 42.

OlfaBensoussia replied:
Taken into consideration

AhmedGrati commented on go/service/contractdefinition/api_op_ListContractDefinitions.go:
```diff
@@ -1,26 +1,47 @@
+	listContractDefinitionsQueryJson, err := json.Marshal(queryInput)
+
+	if err != nil {
+		return nil, fmt.Errorf("unexpected error while marshaling list contract defintions query: %v", err)
```
Same as above.

AhmedGrati commented on go/service/contractdefinition/api_op_ListContractDefinitions.go:
```diff
@@ -1,26 +1,47 @@
 	res, err := c.HTTPClient.Do(req)
 	if err != nil {
-		return nil, sdkErrors.FromError(err).FailedTo(internal.ACTION_HTTP_DO_REQUEST)
+		return nil, fmt.Errorf("error while performing POST request to the endpoint %v: %v", endpoint, err)
```
Also here I recommend using the new error-handling method. Curious to hear your thoughts about it.

AhmedGrati commented on go/service/policies/api_op_ListPolicies.go:
```diff
@@ -7,45 +7,43 @@ import (
+
 	if err != nil {
-		return nil, sdkErrors.FromError(err).FailedTo(internal.ACTION_JSON_MARSHAL)
+		return nil, fmt.Errorf("unexpected error while marshaling list policies query: %v", err)
```
Same as above.

AhmedGrati reviewed (commented):
Good job @OlfaBensoussia 👌, just some small nits and comments about the error handling.
I would recommend also adding some test cases in the unit tests by passing a `QueryInput` in `List` methods of assets, policies, etc...
I invite @NabilHouidi @Nazdroth also to take a look at the PR.

OlfaBensoussia commented:
@AhmedGrati thank you for your reviews again 💯 I addressed all of the comments in my latest commit.

AhmedGrati commented on go/common/apivalidator/validate_query_test.go:
```diff
@@ -0,0 +1,37 @@
+		},
+		"invalid sort order": {
+			sortOrder:  SortOrder("INVALID"),
+			expectedErr:sdkErrors.FailedTo(internal.ACTION_INPUT_VALIDATE),
```
Small nit:
```suggestion
			expectedErr: sdkErrors.FailedTo(internal.ACTION_INPUT_VALIDATE),
```

AhmedGrati reviewed (approved):
It's an LGTM for me, just a small nit.
